File: day01/main.py
def part1(file):
    sample = open(file, 'r')
    first, second = [], []
    for line in sample:
        words = line.split()
        f, s = words[0], words[1]
        first.append(int(f))
        second.append(int(s))

    first.sort()
    second.sort()

    distance = []
    for a, b in zip(first, second):
        d = abs(a - b)
        distance.append(d)
    print('distance for file:',file)
    print(sum(distance))

def part2(file):
    sample = open(file, 'r')
    first =  []
    second = {}
    similarity = []
    for line in sample:
        words = line.split()
        f, s = int(words[0]), int(words[1])
        first.append(f)
        if s not in second:
            second.setdefault(s, 1)
        else:
            second[s] += 1

    for a in first:
        if a in second:
            s = a* second[a]
            similarity.append(s)

    print('similarity for file:',file)
    print(sum(similarity))


part1('sample.txt')
part1('input.txt')


part2('sample.txt')
part2('input.txt')

File: day01/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {

	part1()

	part2()

}

func part2() {
	file, err := os.Open("sample.txt")
	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	file1, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file1.Close()
	sampleSimilarity := calculateSimilarity(file)
	fmt.Println("sample similarity:", sampleSimilarity)

	similarity := calculateSimilarity(file1)
	fmt.Println("similarity:", similarity)
}
func part1() {
	file, err := os.Open("sample.txt")
	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	file1, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file1.Close()
	sampleDist := calculateDistance(file)
	fmt.Println("sample file result:", sampleDist)

	dist := calculateDistance(file1)
	fmt.Println("input file result:", dist)

}

func calculateDistance(file *os.File) int {
	scanner := bufio.NewScanner(file)
	first, second := []int{}, []int{}
	for scanner.Scan() {
		line := scanner.Text()
		numbers := strings.Fields(line)
		if len(numbers) != 2 {
			log.Fatal(errors.New("invalid input"))
		}

		num1, err := strconv.Atoi(numbers[0])
		if err != nil {
			log.Fatal(err)
		}

		num2, err := strconv.Atoi(numbers[1])
		if err != nil {
			log.Fatal(err)
		}
		first = append(first, num1)
		second = append(second, num2)
	}

	slices.Sort(first)
	slices.Sort(second)

	distance := make([]int, len(first))

	for i, _ := range first {
		d := int(math.Abs(float64(second[i] - first[i])))
		distance[i] = d
	}

	res := 0
	for _, n := range distance {
		res += n
	}

	return res
}

func calculateSimilarity(file *os.File) int {
	scanner := bufio.NewScanner(file)
	first := []int{}
	second := make(map[int]int)
	for scanner.Scan() {
		line := scanner.Text()
		numbers := strings.Fields(line)
		if len(numbers) != 2 {
			log.Fatal(errors.New("invalid input"))
		}

		num1, err := strconv.Atoi(numbers[0])
		if err != nil {
			log.Fatal(err)
		}

		num2, err := strconv.Atoi(numbers[1])
		if err != nil {
			log.Fatal(err)
		}

		first = append(first, num1)
		if _, ok := second[num2]; ok {
			second[num2]++
		} else {
			second[num2] = 1
		}

	}
	similarity := 0
	for _, n := range first {
		similarity += n * second[n]
	}

	return similarity
}
